### ISSUE TITLE
dotnet crashes on machine without HOME variable set

### ISSUE THREAD

**Issue by jkotas** [dotnet/cli#8053]
## Steps to reproduce

Run `dotnet new console` on machine without HOME variable set

## Expected  behavior

Works

## Actual behavior

Crash

```
System.ArgumentNullException: Value cannot be null.
Parameter name: path1
   at System.IO.Path.Combine(String path1, String path2)
   at Microsoft.DotNet.Configurer.CliFallbackFolderPathCalculator.get_DotnetUserProfileFolderPath()
   at Microsoft.DotNet.Configurer.FirstTimeUseNoticeSentinel..ctor(CliFallbackFolderPathCalculator cliFallbackFolderPathCalculator)
   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)
   at Microsoft.DotNet.Cli.Program.Main(String[] args)
```

## Environment data
`dotnet --info` output:

## Details

See https://github.com/dotnet/coreclr/pull/15070#issuecomment-345384398:

HOME variable is not set on CI machine. We keep working around it like here: https://github.com/dotnet/coreclr/blob/master/build.sh#L944. It would be much better to fix CLI to not assume that HOME is set. It is pretty likely that customer environments out there do not have HOME set either.

The offending code is here: https://github.com/dotnet/cli/blob/3471d437969ed52d12116f7067b02e6e12a49d03/src/Microsoft.DotNet.Configurer/CliFallbackFolderPathCalculator.cs#L19-L26

**Comment by livarcocc** [comment]
I am not sure what do do when we don't have that set. The CLI has a set of files that it produces and we need a place to put them.

Are you suggesting a more elegant error message? An alternative location?

@emgarten how did you guys handle this in NuGet? I remember you had a similar issue there.

**Comment by justinvp** [comment]
Why not use the following?

```c#
string profileDir = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
```

I believe this should work on both Windows and on Unix when HOME isn't set.

**Comment by livarcocc** [comment]
I wasn't aware of that API 😄. Do you know what it would resolve to?

**Comment by emgarten** [comment]
NuGet throws if HOME isn't set.
https://github.com/NuGet/NuGet.Client/blob/572962f1b1ee890533e5a75163264bd021426dc7/src/NuGet.Core/NuGet.Common/PathUtil/NuGetEnvironment.cs#L226-L239

> Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)

This did not work in the past on the kudu console for azure web sites. See: https://github.com/NuGet/Home/issues/1981

**Comment by justinvp** [comment]
> Do you know what it would resolve to?

On Unix it first tries the HOME environment variable and falls back to calling `getpwuid_r`. If neither of those works it returns a temp path.

https://github.com/dotnet/corefx/blob/0102f69b4274342f5b3d2b6b29827a734369da6d/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L113-L124

https://github.com/dotnet/corefx/blob/bfac45af5dddcbd70b6935fcf92b4b8755f3cd1c/src/Common/src/System/IO/PersistedFiles.Unix.cs#L83

On my mac with HOME unset it returns `/Users/justin`

On Windows it returns `C:\Users\justin`

**Comment by livarcocc** [comment]
Well, there isn't much sense in the CLI working if NuGet will not without HOME being set. restore is required for any meaningful work that we do in the CLI.

I wonder if instead we should just document that having HOME set is a requirement to working with the SDK, if that's not already documented.

**Comment by eerhardt** [comment]
Since using the .NET Core SDK on a build/CI machine is a common scenario, It doesn’t seem like requiring HOME be set is an ideal solution. 

But I also agree that if NuGet doesn’t work, it doesn’t make sense for the rest of the SDK to try to support it. 

Does NuGet require HOME to be set if I override my packages path?  For example using the MSBuild variables, command line —packages, etc?  Overriding packages path to be in the repo is a common pattern in our CI scenarios.

**Comment by livarcocc** [comment]
@emgarten to make sure he sees @eerhardt's question above. That is a good point.

**Comment by emgarten** [comment]
NuGet doesn't require it, it only throws the HOME error message as an alternative to the NRE.

There are a few places where HOME is needed:
* packages folder
* nuget.config user wide settings
* v3 http cache

Potentially nuget reads it too early such as when `--packages` is used before it rules out that the default is needed, but that could be improved.

I agree that requiring HOME is painful, I'd like to see us move away from it also. In the past we've considered adding an env var that points to the `.nuget` folder to override some of these without setting home. We could require that the user override everything individually, but that can be tricky.

**Comment by eerhardt** [comment]
> There are a few places where HOME is needed:
>packages folder
nuget.config user wide settings
v3 http cache

For the nuget.config user-wide settings, if HOME isn't defined I would imagine you treat it the same as if the file doesn't exist.

For the v3 http cache, maybe don't cache anything if HOME isn't defined?

**Comment by emgarten** [comment]
> For the nuget.config user-wide settings, if HOME isn't defined I would imagine you treat it the same as if the file doesn't exist.

We could do this. It would likely need to show warnings since it would be difficult to track down why this isn't working, or why things are so slow due to caching being off.

**Comment by livarcocc** [comment]
Someone else suggested that we use the Environment.SpecialFolder.LocalApplicationData location, which would put us, on my mac, at /Users/<username>/.local/share, which is where NuGet puts its http cache, I believe.

Though, I am not sure it makes sense for us to use that folder for global tools and sentinels. While I can get behind using it for the multijitcore stuff.

Not sure I would want to bifurcate where dotnet things go.

**Comment by peterhuene** [comment]
Given that [NuGet requires `HOME` set for .NET Core](https://github.com/NuGet/NuGet.Client/blob/ea1d1395d0c3dd7b5e22b873cfd71bb8cfb9062a/src/NuGet.Core/NuGet.Common/PathUtil/NuGetEnvironment.cs#L92) without a fallback (even for such basis operations as getting the appropriate string comparer to use for paths, i.e. [GetStringComparerBasedOnOS](https://github.com/NuGet/NuGet.Client/blob/74ff9df4ea3b46c9d0246fa87c863a47e8edb1d2/src/NuGet.Core/NuGet.Common/PathUtil/PathUtility.cs#L21)), we can't fix this on the `dotnet` side alone except for having a much more graceful exit vs. causing an `ArgumentNullException`.

The problem with using `SpecialFolder.UserProfile` and `SpecialFolder.LocalApplicationData` (or any `SpecialFolder` value that uses `HOME` as the basis) is that if `HOME` is not set, CoreFx will use a temporary directory instead.  While it will allow the `dotnet` command to "work", none of the data that was persisted, be it global tools or, assuming NuGet also used these values on .NET Core, packages that were restored or cached, can be shared between `dotnet` invocations.  If CI is the most common scenario where `HOME` is not set, then I don't think this would be acceptable behavior for that scenario.

While CI can certainly set the `HOME` environment variable when invoking `dotnet` (e.g. `HOME=foo dotnet`), it may inadvertantly affect other things than `dotnet`.  Thus, if we were to properly fix this, I'd propose the following:

* Respect a `DOTNET_HOME` environment variable both in `dotnet` and in NuGet (limited to .NET Core is acceptable, since currently there appears to be different behavior for .NET Core already).   This makes it clear that this value is affecting only `dotnet`.  Given that this is close to `DOTNET_ROOT`, other name suggestions are welcomed.
* If `DOTNET_HOME` is not set, fall back to checking `HOME` or `USERPROFILE` based on the platform.
* If the fallback also did not produce a value, fail with an appropriate error message.  For `dotnet`, this might look like:

    ```
    The user profile directory could not be determined. Please use the 'DOTNET_HOME' environment variable to specify the user profile directory.
    ```

@rrelyea is this something that would be acceptable for NuGet?

**Comment by peterhuene** [comment]
We've merged a fix into `release/2.1.4xx` that prevents the crash and now displays a friendly error message asking users to point `DOTNET_CLI_HOME` at the directory where the `.dotnet` sub-directory will be created.  This is akin to `HOME`, except it will only affect `dotnet`.

Note that NuGet still will not function without `HOME` being set, but at least `dotnet` now does.  Ideally NuGet would also respect this variable for .NET Core.  I will open an issue there.
